[B] Add FIREWORK_EXPLODE to the EntityEffect. Adds BUKKIT-4538

LordRalex opened the pull request:
**The Issue:**

Currently, fireworks use a expected time to live number to determine when they need to explode. If you would like to detonate a firework, you would have to set the power level of the firework to 0, but this still leaves a small delay if the firework was very recently launched. Server code will add 1 power to the one provided, multiply that by 10, then add up to 11 more ticks. This causes the firework to only explode at least 10 ticks after being created. This leaves a small chance of the firework not exploding when the plugin wants it to. There is also no way to properly play the effect outside of this randomness.

**Justification for this PR:**

As Minecraft considers a Firework to have an EntityEffect, Bukkit should also follow this and have an EntityEffect.FIREWORK_EXPLODE

**PR Breakdown:**

This PR adds the EntityEffect.FIREWORK_EXPLODE, which is used to play the effect made from a firework.

**Testing Results and Materials:**

A plugin was made that would utilize this new enum value. It would create a new firework on command, set the power to 120, which would mean the firework would expect to detonate at least 1200 ticks later. A scheduled task that was set to run 60 ticks later would call the firework.playEffect(EntityEffect.FIREWORK_EXPLODE) method.

``` java
@Override
    public boolean onCommand(CommandSender sender, Command command, String label, String[] args) {
        if (!(sender instanceof Player)) {
            return true;
        }
        Firework firework = ((Player) sender).getWorld().spawn(((Player) sender).getLocation(), Firework.class);
        FireworkMeta meta = firework.getFireworkMeta();
        meta.setPower(120);
        meta.addEffect(FireworkEffect.builder().with(FireworkEffect.Type.BALL).withColor(Color.WHITE).build());
        firework.setFireworkMeta(meta);
        Bukkit.getScheduler().runTaskLater(this, new Explode(firework), 60);
        return true;
    }

    private class Explode extends BukkitRunnable {

        private final Firework firework;

        public Explode(Firework fire) {
            super();
            firework = fire;
        }

        @Override
        public void run() {
            Bukkit.broadcastMessage("Detonating firework now");
            firework.playEffect(EntityEffect.FIREWORK_EXPLODE);
            firework.remove();
            Bukkit.broadcastMessage("Detonated");
        }
    }
```

Full plugin source: https://dl.dropboxusercontent.com/u/53849969/Bukkit-PRs/src.zip
Compiled binary: https://dl.dropboxusercontent.com/u/53849969/Bukkit-PRs/TestPlugin-1.0-SNAPSHOT.jar

**Relevent PRs**

https://github.com/Bukkit/CraftBukkit/pull/1195 - Original PR made to achieve this, however approach in this PR was more acceptable
https://github.com/Bukkit/Bukkit/pull/900 - Original PR made to achieve this, however approach in this PR was more acceptable

**JIRA Ticket**

https://bukkit.atlassian.net/browse/BUKKIT-4538

gravitylow commented on src/main/java/org/bukkit/EntityEffect.java:
```diff
@@ -43,6 +43,11 @@
+    FIREWORK_EXPLODE(17),
+
+    /**
      * When a sheep eats a LONG_GRASS block.
```
Enums in this class are organized in numerical order, you should follow that by adding this as the last element in the set.

LordRalex replied:
noted and fixed

gravitylow commented:
While I think having a detonate method is nice, I think this implementation is a much better one as it does the same thing with much less of a diff.

LordRalex commented:
That is why I decided this approach would be more acceptable as a solution to the issue, but since others mentioned I should keep both open, I have.

gravitylow commented:
Great. You should squash those commits and force-push.

gravitylow commented on src/main/java/org/bukkit/EntityEffect.java:
```diff
@@ -45,7 +45,12 @@
+    SHEEP_EAT(10),
+    
+    /**
+     * When a firework explodes
```
Needs a period.

Xarinor replied:
Happens to everyone :D

LordRalex replied:
Grammar ftl

LordRalex committed:
Add FIREWORK_EXPLODE to the EntityEffect. Adds BUKKIT-4538

The EntityEffects do not contain the Firework exploding effect. This allows for the playing of the firework effect at will, and even for the repeated playing of the effect. Without this, there is no reliable method to detonate a firework.

LordRalex commented:
Squashed and pushed. (git is scary)

Black-Hole commented:
Maybe you could extend this PR with all of the missing EntityEffects: http://wiki.vg/Protocol#Entity_Status_.280x26.29

LordRalex commented:
Right now this PR is focused on the Firework, as that is the issue it is targeting. It would be better if a new PR was made that would add those.

However if the team would like that to happen, I would not have an issue. I would rather this only fix the issue it was addressed to

gravitylow commented:
Good catch @Black-Hole - it would actually be helpful to have a single commit addressing these missing effects. I don't think that would pose too much of a challenge. Think you could look into that @LordRalex? You'd need to open a different PR.

LordRalex commented:
Not an issue at all. Shall get it made. Would need a leaky ticket to reference it, but otherwise I have no issue.

LordRalex commented:
As requested: https://github.com/Bukkit/Bukkit/pull/907

turt2live commented:
I'm going to go ahead and close this in favour of #907
